// @nrwl/next/src/generators/application/lib/add-linting.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addLinting = void 0;
const tslib_1 = require("tslib");
const linter_1 = require("@nrwl/linter");
const devkit_1 = require("@nrwl/devkit");
const react_1 = require("@nrwl/react");
const run_tasks_in_serial_1 = require("@nrwl/workspace/src/utilities/run-tasks-in-serial");
function addLinting(host, options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const lintTask = yield (0, linter_1.lintProjectGenerator)(host, {
            linter: options.linter,
            project: options.projectName,
            tsConfigPaths: [
                (0, devkit_1.joinPathFragments)(options.appProjectRoot, 'tsconfig.app.json'),
            ],
            unitTestRunner: options.unitTestRunner,
            eslintFilePatterns: [`${options.appProjectRoot}/**/*.{ts,tsx,js,jsx}`],
            skipFormat: true,
        });
        if (options.linter === linter_1.Linter.EsLint) {
            (0, devkit_1.updateJson)(host, (0, devkit_1.joinPathFragments)(options.appProjectRoot, '.eslintrc.json'), (json) => {
                var _a, _b, _c;
                json = (0, react_1.extendReactEslintJson)(json);
                // Turn off @next/next/no-html-link-for-pages since there is an issue with nextjs throwing linting errors
                // TODO(nicholas): remove after Vercel updates nextjs linter to only lint ["*.ts", "*.tsx", "*.js", "*.jsx"]
                json.ignorePatterns = [...json.ignorePatterns, '.next/**/*'];
                json.rules = Object.assign({ '@next/next/no-html-link-for-pages': 'off' }, json.rules);
                // Find the override that handles both TS and JS files.
                const commonOverride = (_a = json.overrides) === null || _a === void 0 ? void 0 : _a.find((o) => ['*.ts', '*.tsx', '*.js', '*.jsx'].every((ext) => o.files.includes(ext)));
                if (commonOverride) {
                    // Only set parserOptions.project if it already exists (defined by options.setParserOptionsProject)
                    if ((_b = commonOverride.parserOptions) === null || _b === void 0 ? void 0 : _b.project) {
                        commonOverride.parserOptions.project = [
                            `${options.appProjectRoot}/tsconfig(.*)?.json`,
                        ];
                    }
                    // Configure custom pages directory for next rule
                    if (commonOverride.rules) {
                        commonOverride.rules = Object.assign(Object.assign({}, commonOverride.rules), { '@next/next/no-html-link-for-pages': [
                                'error',
                                `${options.appProjectRoot}/pages`,
                            ] });
                    }
                }
                (_c = json.extends) !== null && _c !== void 0 ? _c : (json.extends = []);
                if (typeof json.extends === 'string') {
                    json.extends = [json.extends];
                }
                // add next.js configuration
                json.extends.unshift(...['next', 'next/core-web-vitals']);
                // remove nx/react plugin, as it conflicts with the next.js one
                json.extends = json.extends.filter((name) => name !== 'plugin:@nrwl/nx/react');
                json.extends.unshift('plugin:@nrwl/nx/react-typescript');
                if (!json.env) {
                    json.env = {};
                }
                json.env.jest = true;
                return json;
            });
        }
        const installTask = (0, devkit_1.addDependenciesToPackageJson)(host, react_1.extraEslintDependencies.dependencies, react_1.extraEslintDependencies.devDependencies);
        return (0, run_tasks_in_serial_1.runTasksInSerial)(lintTask, installTask);
    });
}
exports.addLinting = addLinting;
